### PR TITLE
New layout - Intrinsic

### PR DESCRIPTION
**Pull request by jpettitt**
Implements #12020

Adds a new layout type "intrinsic" that uses an SVG to size responsive elements thereby giving them an intrinsic size.

cc @Gregable @aghassemi

### CODE REVIEW

**Comment by Gregable** [comment]
@honeybadgerdontcare

File: builtins/amp-img.md [hunk 1]
```diff
@@ -144,6 +144,9 @@ If you want your image to scale as the window is resized but up to a maximum wid
 1. Set `layout=responsive` for `<amp-img>`.
 2. On the container of the image, specify the `max-width:<max width to display image>` CSS attribute.  Why on the container?  An `amp-img` element with `layout=responsive` is a *block-level* element, whereas, `<img>` is *inline*. Alternatively, you could set `display: inline-block` in your CSS for the amp-img element.
 
+#### The difference between responsive and intrinsic layout
+
+Both the `responsive` and `intrinsic` layouts create an image that will scale automatically.  The main difference is that the `intrinsic` layout uses an SVG image as it's scaling element.  This will make it behave in the same way as a standard html image while retaining the benifit of the browser knowing the image size on initial layout. The `intrinsic` layout will have an intrinsic size and will inflate a floated `div` until it reaches either the natural image size or a CSS constraint like `max-width`. The `responsive` layout will render 0x0 in a floated `div` becasue it takes its size from the parent which in a float has no natural size.
```
Comment (honeybadgerdontcare) [inline builtins/amp-img.md]: s/benifit/benefit

File: extensions/amp-carousel/validator-amp-carousel.protoascii [hunk 2]
```diff
@@ -64,5 +64,6 @@ tags: {  # <amp-carousel>
     supported_layouts: FLEX_ITEM
     supported_layouts: NODISPLAY
     supported_layouts: RESPONSIVE
+    supported_layouts: INTRINSIC
```
Comment (honeybadgerdontcare) [inline extensions/amp-carousel/validator-amp-carousel.protoascii]: place after `FLEX_ITEM`

**Comment by honeybadgerdontcare** [comment]
@jpettitt I think you could move forward with this, just remove the validator files. I'll get the validator engine changes in and put the protoascii changes in at the same time.

File: extensions/amp-carousel/validator-amp-carousel.protoascii [hunk 3]
```diff
@@ -66,5 +66,6 @@ tags: {  # <amp-carousel>
     supported_layouts: INTRINSIC
     supported_layouts: NODISPLAY
     supported_layouts: RESPONSIVE
+    supported_layouts: INTRINSIC
```
Comment (aghassemi) [inline extensions/amp-carousel/validator-amp-carousel.protoascii]: it is already added above.

File: extensions/amp-ad/validator-amp-ad.protoascii [hunk 4]
```diff
@@ -75,6 +75,7 @@ tags: {  # <amp-ad>
     supported_layouts: INTRINSIC
     supported_layouts: NODISPLAY
     supported_layouts: RESPONSIVE
+    supported_layouts: INTRINSIC
```
Comment (aghassemi) [inline extensions/amp-ad/validator-amp-ad.protoascii]: rule is already there for both of these

File: spec/amp-html-layout.md [hunk 5]
```diff
@@ -120,6 +120,10 @@ Supported values for the `layout` attribute:
       <td><code>flex-item</code></td>
       <td>The element and other elements in its parent with layout type <code>flex-item</code> take the parent container's remaining space when the parent is a flexible container (i.e., <code>display: flex</code>). The <code>width</code> and <code>height</code> attributes are not required.</td>
     </tr>
+    <tr>
+      <td><code>intrinsic</code></td>
+      <td>The element takes the space available to it and resizes its height automatically to the aspect ratio given by the <code>width</code> and <code>height</code> attributes. This layout works very well for most AMP elements, including <code>amp-img</code>, <code>amp-video</code>, etc.  The available space depends on the parent element and can also be customized using <code>max-width</code> CSS. The <code>width</code> and <code>height</code> attributes must be present.  This layout differs from <code>responsive</code> by having an intrinsic height and width. This is most apparent inside a floated element where a <code>responsive</code> layout will render 0x0 and <code>intrinsic</code> layout will inflate to the smaller of it's natural size or any CSS constraint. </td>
```
Comment (aghassemi) [inline spec/amp-html-layout.md]: s/it's natural/its natural

File: spec/amp-html-layout.md [hunk 6]
```diff
@@ -120,6 +120,10 @@ Supported values for the `layout` attribute:
       <td><code>flex-item</code></td>
       <td>The element and other elements in its parent with layout type <code>flex-item</code> take the parent container's remaining space when the parent is a flexible container (i.e., <code>display: flex</code>). The <code>width</code> and <code>height</code> attributes are not required.</td>
     </tr>
+    <tr>
+      <td><code>intrinsic</code></td>
+      <td>The element takes the space available to it and resizes its height automatically to the aspect ratio given by the <code>width</code> and <code>height</code> attributes. This layout works very well for most AMP elements, including <code>amp-img</code>, <code>amp-video</code>, etc.  The available space depends on the parent element and can also be customized using <code>max-width</code> CSS. The <code>width</code> and <code>height</code> attributes must be present.  This layout differs from <code>responsive</code> by having an intrinsic height and width. This is most apparent inside a floated element where a <code>responsive</code> layout will render 0x0 and <code>intrinsic</code> layout will inflate to the smaller of its natural size or any CSS constraint. </td>
+    </tr>
```
Comment (bpaduch) [inline spec/amp-html-layout.md]: You should also mention `intrinsic` in the `Summary of Layout Requirements & Behaviors` table that's at the bottom of this doc.

File: src/layout.js [hunk 7]
```diff
@@ -424,6 +429,18 @@ export function applyStaticLayout(element) {
...
     element.sizerElement = sizer;
+  } else if (layout == Layout.INTRINSIC) {
+    // Intrinsic uses an svg inside the sizer element rather than the padding trick
+    // Note a naked svg won't work becasue other thing expect the i-amphtml-sizer element
+    const sizer = element.ownerDocument.createElement('i-amphtml-sizer');
+    const intrinsicSizer = element.ownerDocument.createElement('img');
+    sizer.classList.add('i-amphtml-sizer');
+    intrinsicSizer.classList.add('i-amphtml-intrinsic-sizer');
+    intrinsicSizer.setAttribute('src',
+        `data:image/svg+xml;charset=utf-8,<svg height="${height}" width="${width}" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1"></svg>`);
```
Comment (jridgewell) [inline src/layout.js]: Nit: are the `xmlns:xlink` and `version` attributes necessary?

**Pull request merged**